Firestore/Swift/Source/Codable/ServerTimestamp.swift

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import FirebaseFirestore
+
+#if swift(>=5.1)
+
+  /// A marker protocol that indicates the type can be annotated with an
+  /// `@ServerTimestamp` property wrapper. Only `Timestamp`, `Date`, and `NSDate`
+  /// can be wrapped with `@ServerTimestamp`.
+  public protocol ServerTimestampWrappable {}
+
+  extension Date: ServerTimestampWrappable {}
+  extension NSDate: ServerTimestampWrappable {}
+  extension Timestamp: ServerTimestampWrappable {}
```

Comment:
@ryanwilson I'm mostly interested in your take on this kind of type constraint. The rest of this is essentially just like ExplicitNull so feel free to ignore.

Reply:
I think it's reasonable, and looks like a good solution. One thing that may be helpful is to actually encode what we're trying to get out of the type and implement it in the extension, making the decode a bit cleaner. Potentially just having two properties: seconds and nanoseconds. What do you think?

Reply:
That's a great idea. It makes the mechanism completely extensible. I've added a test showing how @ServerTimestamp apply to arbitrary types in 4121604.

Reply:
Looks great!